File: src/daemon.ts
import { spawn } from "node:child_process";
import fs from "node:fs/promises";
import fsSync from "node:fs";
import net from "node:net";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { assertConfigured, loadBridgeConfig, resolveApiKey } from "./config.js";
import { getBridgeConfigPath, getBridgeHome, getDaemonLogPath, getDaemonStatePath } from "./paths.js";
import type { DaemonState, DaemonStatus, StartDaemonResult, StopDaemonResult } from "./types.js";

function bridgeUrl(host: string, port: number): string {
  return `http://${host}:${port}/v1`;
}

async function fileExists(filePath: string): Promise<boolean> {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

function isProcessRunning(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch {
    return false;
  }
}

export async function canConnect(host: string, port: number, timeoutMs = 800): Promise<boolean> {
  return new Promise((resolve) => {
    const socket = net.createConnection({ host, port });
    socket.setTimeout(timeoutMs);
    socket.once("connect", () => {
      socket.destroy();
      resolve(true);
    });
    socket.once("timeout", () => {
      socket.destroy();
      resolve(false);
    });
    socket.once("error", () => resolve(false));
  });
}

async function readDaemonState(statePath: string): Promise<DaemonState | null> {
  try {
    return JSON.parse(await fs.readFile(statePath, "utf8")) as DaemonState;
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") {
      return null;
    }
    throw error;
  }
}

export async function daemonStatus(bridgeHome = getBridgeHome()): Promise<DaemonStatus> {
  const statePath = getDaemonStatePath(bridgeHome);
  const logPath = getDaemonLogPath(bridgeHome);
  const state = await readDaemonState(statePath);
  const running = state ? isProcessRunning(state.pid) : false;
  const stale = Boolean(state && !running);

  return {
    statePath,
    logPath: state?.logPath ?? logPath,
    running,
    pid: state?.pid,
    stale
  };
}

export async function startDaemon(bridgeHome = getBridgeHome()): Promise<StartDaemonResult> {
  const configPath = getBridgeConfigPath({ env: { ...process.env, CODEX_PROVIDER_BRIDGE_HOME: bridgeHome } });
  const config = await loadBridgeConfig(configPath);
  assertConfigured(config);

  const key = resolveApiKey(config);
  if (!key.apiKey) {
    throw new Error(`Missing API key. Run "codex-provider-bridge setup" or set ${config.apiKeyEnv}.`);
  }

  const status = await daemonStatus(bridgeHome);
  const url = bridgeUrl(config.host, config.port);
  if (status.running) {
    return { alreadyRunning: true, pid: status.pid, statePath: status.statePath, logPath: status.logPath, url };
  }
  if (status.stale) {
    await fs.rm(status.statePath, { force: true });
  }

  if (await canConnect(config.host, config.port)) {
    throw new Error(`本地端口 ${config.port} 已被占用。请运行 codex-provider-bridge status 查看，或运行 setup 换一个端口。`);
  }

  await fs.mkdir(bridgeHome, { recursive: true });
  const logPath = getDaemonLogPath(bridgeHome);
  const out = fsSync.openSync(logPath, "a");
  const err = fsSync.openSync(logPath, "a");
  const cliPath = fileURLToPath(import.meta.url).replace(/[\\/]daemon\.js$/, `${path.sep}cli.js`);
  let child;
  try {
    child = spawn(process.execPath, [cliPath, "serve"], {
      cwd: process.cwd(),
      detached: true,
      env: {
        ...process.env,
        CODEX_PROVIDER_BRIDGE_HOME: bridgeHome
      },
      stdio: ["ignore", out, err]
    });
  } finally {
    fsSync.closeSync(out);
    fsSync.closeSync(err);
  }

  child.unref();
  if (!child.pid) {
    throw new Error("后台进程启动失败：没有拿到进程 ID。");
  }

  const state: DaemonState = {
    pid: child.pid,
    startedAt: new Date().toISOString(),
    command: `${process.execPath} ${cliPath} serve`,
    logPath,
    host: config.host,
    port: config.port
  };
  await fs.writeFile(status.statePath, `${JSON.stringify(state, null, 2)}\n`, "utf8");

  return {
    alreadyRunning: false,
    pid: child.pid,
    statePath: status.statePath,
    logPath,
    url
  };
}

export async function stopDaemon(bridgeHome = getBridgeHome()): Promise<StopDaemonResult> {
  const status = await daemonStatus(bridgeHome);
  if (!status.pid) {
    return { stopped: false, wasRunning: false, statePath: status.statePath };
  }

  if (!status.running) {
    await fs.rm(status.statePath, { force: true });
    return { stopped: false, wasRunning: false, pid: status.pid, statePath: status.statePath };
  }

  process.kill(status.pid, "SIGTERM");
  for (let attempt = 0; attempt < 20; attempt += 1) {
    await new Promise((resolve) => setTimeout(resolve, 100));
    if (!isProcessRunning(status.pid)) {
      break;
    }
  }

  await fs.rm(status.statePath, { force: true });
  return { stopped: true, wasRunning: true, pid: status.pid, statePath: status.statePath };
}

export async function hasDaemonState(bridgeHome = getBridgeHome()): Promise<boolean> {
  return fileExists(getDaemonStatePath(bridgeHome));
}

File: src/daemon.rs
use std::fs::OpenOptions;
use std::path::PathBuf;
use std::process::{Command, Stdio};
use std::time::Duration;

use chrono::Utc;
use tokio::fs;
use tokio::net::TcpStream;
use tokio::time::timeout;

use crate::config::{assert_configured, load_bridge_config, resolve_api_key};
use crate::error::{err, Result};
use crate::paths::{
    get_bridge_config_path, get_bridge_home, get_daemon_log_path, get_daemon_state_path,
};
use crate::types::{DaemonState, DaemonStatus, StartDaemonResult, StopDaemonResult};

pub async fn can_connect(host: &str, port: u16, timeout_ms: u64) -> bool {
    let target = format!("{host}:{port}");
    matches!(
        timeout(
            Duration::from_millis(timeout_ms),
            TcpStream::connect(target)
        )
        .await,
        Ok(Ok(_))
    )
}

pub async fn daemon_status(bridge_home: Option<PathBuf>) -> Result<DaemonStatus> {
    let bridge_home = match bridge_home {
        Some(path) => path,
        None => get_bridge_home(None)?,
    };
    let state_path = get_daemon_state_path(&bridge_home);
    let default_log_path = get_daemon_log_path(&bridge_home);
    let state = read_daemon_state(&state_path).await?;
    let running = state
        .as_ref()
        .is_some_and(|state| is_process_running(state.pid));
    Ok(DaemonStatus {
        state_path,
        log_path: state
            .as_ref()
            .map(|state| state.log_path.clone())
            .unwrap_or(default_log_path),
        running,
        pid: state.as_ref().map(|state| state.pid),
        stale: state.is_some() && !running,
    })
}

pub async fn start_daemon() -> Result<StartDaemonResult> {
    let bridge_home = get_bridge_home(None)?;
    let config_path = bridge_home.join("config.json");
    let config = load_bridge_config(Some(&config_path)).await?;
    assert_configured(&config)?;
    if resolve_api_key(&config).api_key.is_none() {
        return Err(err(format!(
            "Missing API key. Run \"codex-provider-bridge setup\" or set {}.",
            config.api_key_env
        )));
    }

    let status = daemon_status(Some(bridge_home.clone())).await?;
    let url = format!("http://{}:{}/v1", config.host, config.port);
    if status.running {
        return Ok(StartDaemonResult {
            already_running: true,
            pid: status.pid,
            state_path: status.state_path,
            log_path: status.log_path,
            url,
        });
    }
    if status.stale {
        let _ = fs::remove_file(&status.state_path).await;
    }
    if can_connect(&config.host, config.port, 800).await {
        return Err(err(format!(
            "本地端口 {} 已被占用。请运行 codex-provider-bridge status 查看，或运行 setup 换一个端口。",
            config.port
        )));
    }

    fs::create_dir_all(&bridge_home).await?;
    let log_path = get_daemon_log_path(&bridge_home);
    let stdout = OpenOptions::new()
        .create(true)
        .append(true)
        .open(&log_path)?;
    let stderr = OpenOptions::new()
        .create(true)
        .append(true)
        .open(&log_path)?;
    let exe = std::env::current_exe()?;
    let mut command = Command::new(&exe);
    command
        .arg("serve")
        .env("CODEX_PROVIDER_BRIDGE_HOME", &bridge_home)
        .stdin(Stdio::null())
        .stdout(Stdio::from(stdout))
        .stderr(Stdio::from(stderr));
    #[cfg(windows)]
    {
        use std::os::windows::process::CommandExt;
        command.creation_flags(windows_sys::Win32::System::Threading::CREATE_NO_WINDOW);
    }
    let child = command.spawn()?;
    let pid = child.id();
    let state = DaemonState {
        pid,
        started_at: Utc::now().to_rfc3339(),
        command: format!("{} serve", exe.display()),
        log_path: log_path.clone(),
        host: config.host,
        port: config.port,
    };
    fs::write(
        &status.state_path,
        format!("{}\n", serde_json::to_string_pretty(&state)?),
    )
    .await?;
    Ok(StartDaemonResult {
        already_running: false,
        pid: Some(pid),
        state_path: status.state_path,
        log_path,
        url,
    })
}

pub async fn stop_daemon() -> Result<StopDaemonResult> {
    let status = daemon_status(None).await?;
    let Some(pid) = status.pid else {
        return Ok(StopDaemonResult {
            stopped: false,
            pid: None,
            state_path: status.state_path,
        });
    };
    if !status.running {
        let _ = fs::remove_file(&status.state_path).await;
        return Ok(StopDaemonResult {
            stopped: false,
            pid: Some(pid),
            state_path: status.state_path,
        });
    }

    terminate_process(pid);
    for _ in 0..20 {
        tokio::time::sleep(Duration::from_millis(100)).await;
        if !is_process_running(pid) {
            break;
        }
    }
    let _ = fs::remove_file(&status.state_path).await;
    Ok(StopDaemonResult {
        stopped: true,
        pid: Some(pid),
        state_path: status.state_path,
    })
}

async fn read_daemon_state(state_path: &std::path::Path) -> Result<Option<DaemonState>> {
    match fs::read_to_string(state_path).await {
        Ok(raw) => Ok(Some(serde_json::from_str(&raw)?)),
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(None),
        Err(error) => Err(error.into()),
    }
}

fn is_process_running(pid: u32) -> bool {
    #[cfg(windows)]
    {
        Command::new("cmd")
            .args([
                "/C",
                &format!("tasklist /FI \"PID eq {pid}\" | findstr /R \"\\<{pid}\\>\""),
            ])
            .status()
            .map(|status| status.success())
            .unwrap_or(false)
    }
    #[cfg(not(windows))]
    {
        Command::new("kill")
            .args(["-0", &pid.to_string()])
            .status()
            .map(|status| status.success())
            .unwrap_or(false)
    }
}

fn terminate_process(pid: u32) {
    #[cfg(windows)]
    {
        let _ = Command::new("taskkill")
            .args(["/PID", &pid.to_string(), "/T"])
            .status();
    }
    #[cfg(not(windows))]
    {
        let _ = Command::new("kill")
            .args(["-TERM", &pid.to_string()])
            .status();
    }
}

#[allow(dead_code)]
fn _config_path_for_bridge_home(bridge_home: &std::path::Path) -> PathBuf {
    let mut env = std::collections::HashMap::new();
    env.insert(
        "CODEX_PROVIDER_BRIDGE_HOME".to_string(),
        bridge_home.display().to_string(),
    );
    get_bridge_config_path(Some(&env)).expect("valid path")
}
